### PR TITLE
Prepare support for Dynamic Instrumentation

### PR DESCRIPTION
**Pull request by watson**
Prepare the Node.js tracer to be used with the [Dynamic Instrumentation product](https://docs.datadoghq.com/dynamic_instrumentation/) (DI).

Please note, that even after this PR has been merged, it's not possible to use DI with the Node.js tracer, as 1) this isn't the full implementation and 2) the Datadog UI hasn't been updated to support DI for Node.js applications (it's currently behind a feature flag that needs to be enabled by Datadog).

### Implementation details

Probes are added as breakpoints using the [V8 Inspector Protocol](https://v8.dev/docs/inspector). The inspector runs in a worker thread and will pause the main thread temporarily while gathering the required information related to the probe (as of this PR no information is gathered, but this is a required step once we start to gather the local state snapshot).

### Included in this PR

- Support for line based log probes
- Support for loading probes via Remote Configuration
- Support for sending probe status to the debugger-backend
- Support for sending log data to debugger-backend

### Out of scope for this PR

- No support for function probes
- No support for Metric/Span/Span Tag probes
- No support for sampling
- No support for capturing of snapshots
- No support for templates (in either conditions or messages)
- No support for attaching span/trace ID to payload

### Notes to reviewers

To enable, set `DD_DYNAMIC_INSTRUMENTATION_ENABLED=true` and enable the `live_debugger_language_js` feature flag in the UI.

### Tasks

- [x] Add tests
- [x] Go over remaining TODO comments and resolve any that needs resolving before merging
- [x] Either trigger `test:debugger` during CI, or change to use the Mocha test-runner

### CODE REVIEW

**Comment by watson** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-js/4492?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4492** <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-js/4492?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4521** <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-js/4521?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#4520](https://github.com/DataDog/dd-trace-js/pull/4520) <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-js/4520?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#4515** <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-js/4515?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @watson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by rochdev** [comment]
It would make assigning a code owner easier if this was its own folder directly under `packages` (which we could also decide to actually release as its own package later on).

File: CODEOWNERS
Comment (simon-id) [inline CODEOWNERS:7]: if i were you i would do this instead, otherwise you're never gonna merge anything haha
```suggestion
/packages/dd-trace/src/debugger/ @DataDog/debugger @Datadog/dd-trace-js
/integration-tests/debugger/ @DataDog/debugger @Datadog/dd-trace-js
```

Comment (watson) [reply CODEOWNERS:7]: I thought the first line in the file automatically added our team to all files by using `*`?

File: packages/dd-trace/src/debugger/devtools_client/status.js
Comment (shatzi) [inline packages/dd-trace/src/debugger/devtools_client/status.js:27]: I think there are two features missing from probe status implementation:
1. don't repeat status messages. for example, only send emitting/error once.
2. once every hour, send probe status for all probes on the list. this should basiclly emit the last status of that probe.

If you ask why we are doing 2 - well, that because the UI would like to query the events from the last hour and get all the probe status - even if they where sent 3 days ago.

File: packages/dd-trace/src/debugger/index.js
Comment (shatzi) [inline packages/dd-trace/src/debugger/index.js:24]: going back to the race condition - I see that the postMessage happen without waiting for the previous one to complete. is that right?

File: packages/dd-trace/src/debugger/index.js
Comment (shatzi) [inline packages/dd-trace/src/debugger/index.js:52]: what is the state of process if that happens? what if a breakpoint hit and our worker thread is dead?

**Review by shatzi — COMMENTED** [review]
great progress!

File: CODEOWNERS
Comment (simon-id) [reply CODEOWNERS:7]: I don't think so no, you can double check, but I remember being in this situation and blocking myself from merging anything, but i could be wrong

Comment (watson) [reply CODEOWNERS:7]: I've just verified, and it works like I expected 😀

Take a look at this line in `CODEOWNERS`:
https://github.com/DataDog/dd-trace-js/blob/99497961d3dc70b5b55eaef3d52248f93936f0ab/CODEOWNERS#L13

Here's [a 3 week old PR](https://github.com/DataDog/dd-trace-js/pull/4609) that changes a file under that directory. If you see what happened once the PR was marked as "ready for review", both `@DataDog/ci-app-libraries` and `@DataDog/dd-trace-js` was requested for a code owners review:

image

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:61]: > If I understand this right, we sending Received probe status when unapply a probe?

No it's the opposite. If the `action` is anything else than `unapply`, _then_ we call `ackReceived()`

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:80]: Ah, good point. If you don't mind, I'll leave this for a upcoming PR to not complicate this PR further. I'll make an issue for it.

File: CODEOWNERS
Comment (simon-id) [reply CODEOWNERS:7]: Then what is the point of this line for example then: https://github.com/DataDog/dd-trace-js/blob/99497961d3dc70b5b55eaef3d52248f93936f0ab/CODEOWNERS#L8

File: packages/dd-trace/src/debugger/devtools_client/send.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/send.js:18]: > the thread should be the thread where the snapshot was captured, not the thread that is sending

I thought the idea of the `logger` object was to capture information about the logger itself and not the code that was being instrumented. But if that's not the case, then I think there's a few more issues with this `logger` object: I imagine both `name` and `method` should also updated to point to the code being instrumented and not this file?

File: packages/dd-trace/src/debugger/devtools_client/state.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/state.js:17]: Ups, thanks

File: packages/dd-trace/src/debugger/devtools_client/state.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/state.js:42]: Yes it does, but thanks for addressing it ([source](https://chromedevtools.github.io/devtools-protocol/v8/Debugger/#event-scriptParsed))

File: packages/dd-trace/src/debugger/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/index.js:24]: > going back to the race condition

Which race condition? I think I've mentioned several. But to my knowledge I've dealt with all of them, so there should be none left that I know of.

> I see that the postMessage happen without waiting for the previous one to complete. is that right?

That's correct. I don't think there's an issue with that. Have you spotted anything that I don't see?

File: packages/dd-trace/src/debugger/devtools_client/status.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/status.js:27]: > 1. don't repeat status messages. for example, only send emitting/error once.

So if a probe is first created and then later modified, we should only send the `RECEIVED` payload once? How will the UI then know that the RC message to modify the probe was received?

So for a givne unique probe id, we never want to send the same diagnostic message more than once? E.g. if a probe is first created and then later modified, we should only send the `RECEIVED` payload once? And if the same probe is being triggered multiple times, we should also only send the `EMITTING` payload the first time it was triggered?

> 2. once every hour, send probe status for all probes on the list. this should basiclly emit the last status of that probe.
>
> If you ask why we are doing 2 - well, that because the UI would like to query the events from the last hour and get all the probe status - even if they where sent 3 days ago.

I think this state management should really be handled by the backend database and not delegated to the tracers - But I have a feeling you agree with me already 😅 In any case, I'll implement this and then try to forget about it as soon as possible 😭

Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/status.js:27]: Follow up question: If a given probe experiences more than one unique error, we still only want to send the first one we see within a given hour? So if we in hour one, see 3 errors, all unique, we only send the first one, then one hour and one second after we saw the first error, a new error happens, which we sent, effectively dropping the two inbetween?

Comment (shatzi) [reply packages/dd-trace/src/debugger/devtools_client/status.js:27]: Yeah, you are right - we send a single probe per probe-id+version combo. so when a probe get updated, the probe-version increase by 1 and then you emit new RECEIVED message. then that status can be upgraded to INSTALLED and EMITTING (or ERROR/BLOCKED on the sad path).

Yup. the backend can store that at the backend - but that train left station long time ago

File: CODEOWNERS
Comment (watson) [reply CODEOWNERS:7]: No need to have the `@Datadog/dd-trace-js` team on that line

Comment (watson) [reply CODEOWNERS:7]: Nope, you're actually 100% correct. I just discovered a flaw in my logic above. The PR I linked to happened to also touch 3 files that was directly owned by `@Datadog/dd-trace-js`, hence why the PR requested a code owners review from both teams 🤦. My bad. I'll update the entry

Comment (watson) [reply CODEOWNERS:7]: Also, I managed to finally track down the docs for how the `*` entry on the first line of the file should be interpreted:

```
# These owners will be the default owners for everything in
# the repo. Unless a later match takes precedence,
# @global-owner1 and @global-owner2 will be requested for
# review when someone opens a pull request.
*       @global-owner1 @global-owner2
```

_([source](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file))_

Comment (simon-id) [reply CODEOWNERS:7]: Feel free to close that thread when it's fixed :+1:

File: packages/dd-trace/src/debugger/devtools_client/send.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/send.js]: Fixed

File: packages/dd-trace/src/debugger/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/index.js]: I gave it some extra thought after our Zoom call and decided to simplify the listener body, since any exit of the worker thread shouldn't happen and should be treated as an error - no matter the exit code. I've updated the code accordingly.

File: packages/dd-trace/src/debugger/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/index.js:24]: We had a chat about this over Zoom and I now understand the race condition you're referring to:

If, say, there's just a single breakpoint active and an RC request comes in to remove that breakpoint, after removing it, we post `Debugger.disable` on the inspector protocol session since there's no reason to keep the debugger session active if there's no active breakpoints.

However, if while we process the request to remove the last breakpoint, but before disabling the inspector session, get a new RC request to add a new breakpoint, we will not post `Debugger.enable` because we see that the session is already active.

This is the race condition.

File: packages/dd-trace/src/debugger/devtools_client/status.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/status.js:36]: I've now implemented a check to ensure we don't send the same acknowledge-request twice within the same hour. I have specifically _not_ implemented a timer to resubmit the status of all probes ones an hour as this is no longer strictly needed according to @cimi. I've added this topic to the agent sync agenda backlog.

File: packages/dd-trace/src/debugger/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/index.js:24]: For now I added a mutex lock. I'll revet that change if we decided to just never tear it down

**Review by bengl — CHANGES_REQUESTED** [review]

**Review by Qard — COMMENTED** [review]
I'll finish reviewing tomorrow, but a bit of feedback for now:

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:52]: I'm not sure I understand... what's the benefit of doing that?

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:24]: 🤷 

> What does the spec for this say?

Ha, if there just was a spec that would be lovely. I haven't been able to find any documentation or spec that covers this unfortunately.

Calling `require('node:worker_threads').threadId` from the main thread always gives you `0` which is not the actual thread ID in the Linux sense of the word. So I don't think that's a meaningful number either. Also, I haven't found a way to get anything resembling a "thread name" from Node.js. I'm not sure we even have the concept in Node.js. 

> If the instrumented app is on the main thread, PID will be incorrect as the TID. Maybe that's okay? Depend what the spec here says.

That was my fallback since I couldn't find any better number here. But I'm aware it's not going to be correct if the instrumented process _isn't_ the main thread.

File: packages/dd-trace/src/debugger/devtools_client/inspector_promises_polyfill.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/inspector_promises_polyfill.js:14]: Probably. However, I just blindly lifted the code from Node.js core (source linked in the code comment above) and on purpose didn't try to clean it up. I think that's always the best solution as it makes it easier to update the code in the future if it changes up-stream

File: packages/dd-trace/src/debugger/devtools_client/state.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/state.js:35]: `eval()`/`new Function()` doesn't trigger the `Debugger.scriptParsed` event - I guess that's considered part of the containing script file on disk.

`setTimeout(() => {}, ...)` doesn't trigger  the `Debugger.scriptParsed` event either (and Node.js doesn't support parsing a string to `setTimeout` like the browser does

A dynamic / delayed require however does trigger the `Debugger.scriptParsed` event as we stipulated on the call yesterday, so I do need to add support for that.

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:75]: > It's worth leaving a comment as to what this lock is for. I think it's because you want to guarantee some ordering? 

It's to fix a race condition discussed here: https://github.com/DataDog/dd-trace-js/pull/4492#discussion_r1750071683

I'll add a comment 👍 

> At any rate, you don't really need a lock for the default/error case.

You're right, but I thought the code looked cleaner if I just wrapped the entire switch statement in the lock instead of having to set and release the lock inside each `case` block.

Also, I missed adding a `try { ... } finally { release() }` block around the switch statement below to ensure that the lock gets released if anything in-between threw. I'm going to fix that.

But please let me know if you think it could be done prettier / better in another way ☺️

@Qard Do you still think a queue would be better based on what I just wrote?

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:34]: The only time more than one probe is hit at the same time, it because there's multiple breakpoints in the same location - in which case the `location.file` property is going to be identical for each of the hit probes. So we can safely just use the first one in the array to populate the location `name` field.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:11]: Good point. I'll improve the code

Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:11]: Actually, on second thought, I actually considered this and the code is like this on purpose: You're right that something like an `unhandledRejection` will be happen (not sure if it's exactly that or something similar), but in any case, that will just terminate the worker thread and it will be logged via the `error` listener in `packages/dd-trace/src/debugger/index.js`.

In general I believe there's two philosophies to error handling in a case like this: Either try to guard against any unforeseen error, or just accept that the code might have a bug and crash/fail as early as possible. The latter makes the code much more readable / maintainable IMO.

Of course we shouldn't bring down the customer app, but in this case, all that will happen is that the thread will crash, an error be reported, and Dynamic Instrumentation would just stop working. We probably couldn't do any better if we tried to handle it all in-thread as any error that might exist, would probably not be "fixed" by catching it and would persist next time the event listener was called.

I've verified, that if the worker tread crashes while the main thread is paused, the debugger is going to be detached and the main thread will resume running.

One improvement however that I should do, is to call `rc.removeProductHandler('LIVE_DEBUGGING')` if the worker thread ever exists. That way RC would also know that we're no longer able to listen/handle any RC updates.

So based on all that, I vote for just leaving the code as is. What's your thoughts?

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:66]: This is handled by the caller to `processMsg` which in this case wraps the call to `processMsg` inside a try-catch block.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js]: Fixed 👍

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js]: Fixed 👍

File: packages/dd-trace/src/debugger/devtools_client/inspector_promises_polyfill.js
Comment (bengl) [reply packages/dd-trace/src/debugger/devtools_client/inspector_promises_polyfill.js:14]: Yeah this seems like no big deal, let's just leave it as is.

File: packages/dd-trace/src/debugger/devtools_client/state.js
Comment (bengl) [reply packages/dd-trace/src/debugger/devtools_client/state.js:35]: Let's not worry about this for now. We're using the inspector protocol, so expectations around this should be the same as in the Chrome debugger.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (bengl) [reply packages/dd-trace/src/debugger/devtools_client/index.js:33]: Ok, but that's not necessarily clear at first glance, so it warrants a comment to clarify the assumption.

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (bengl) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:67]: All good then.

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:53]: Post zoom call update: I understand now... returning instead of awaiting will mean fewer promises are created. I'll update the code

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:35]: Add a comment 👍

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js]: Updated the code to use `return` instead 👍

File: packages/dd-trace/src/debugger/devtools_client/config.js
Comment (rochdev) [inline packages/dd-trace/src/debugger/devtools_client/config.js:19]: It can be `unix:` for Unix domain sockets, in which case you'd get a pathname and no host.

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (rochdev) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: I think this is easy enough to do without a library that we should do it without. Adding libraries, especially ones that are unmaintained like this one, can add arrays of problems and also impact the whole library and not just DI.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: @rochdev I can implement the lock manually like so:

```js
async function lock () {
  if (lock.p) await lock.p
  let resolve
  lock.p = new Promise((_resolve) => { resolve = _resolve }).then(() => { lock.p = null })
  return resolve
}
```

But there's something to be said about well tested small npm modules to encapsulate complexity and reduce risk of bugs. If I read the code I just posted above, I couldn't say off the top of my head if it would work as expected. I would have to test it, which is what the npm module already does for me.

It's a balance of course. We don't want a million dependencies, but we also don't want to reinvent the well if reinventing comes with added risk (unforeseen bugs in this case).

Btw, to my knowledge @mafintosh is still maintaining `mutexify`. It's just done, unless there's a bug to be fixed - hence why it hasn't received many updates over the last years.

However, this isn't a hill I'm willing to die on 😄 So I can update the code as you suggested, but I'd love to hear if @bengl or @Qard had some input first?

Comment (Qard) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: Yeah, I would go for a task queue instead. I think it's clearer that way what is intended. Could be as simple as:

```js
let lastTask = Promise.resolve()
function queueTask(taskFn) {
  return (lastTask = last.then(taskFn))
}
```

Comment (mafintosh) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: image

Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: @Qard I don't think the code your posted handles errors that might occur inside `taskFn` 🤔 At least, after implementing it, it looks like errors are escaping the outer try-catch block. I've spent a few hours on it now trying to get it to work properly in all edge cases (which is exactly my point about using an existing npm module 😉)

File: packages/dd-trace/src/debugger/devtools_client/config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/config.js]: Fixed 👍

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/remote_config.js:86]: Btw, I added a new commit that contains an test which fails if the race condition is triggered

File: packages/dd-trace/src/debugger/devtools_client/config.js
Comment (OmerRaviv) [reply packages/dd-trace/src/debugger/devtools_client/config.js:1]: No strong preference from my end - as long as we are consistent with customer-facing things like environment variables, I'm happy. In this case, I don't feel strongly whether consistency within a given tracer or consistency across the different tracers should win.

Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/config.js:1]: After having discussed this with @rochdev on Slack, I'm going to leave the folder name as is for now. I can always be changed later if we think it's required. The important thing is that it doesn't affect the name of the environment variables or any other customer facing name/value.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js]: I've updated the content of the variables now based on discussion with the DI team.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (uurien) [inline packages/dd-trace/src/debugger/devtools_client/index.js:52]: how big can be the `probes` array? Just to keep in mind that there is a max of request that the tracer could do:
https://github.com/DataDog/dd-trace-js/blob/bd2f8b2b0c0cd0d35a600912b99aa1c3f7e74d6d/packages/dd-trace/src/exporters/common/request.js#L18
if `probes.length` could be bigger than 8, max of first 8 will be sent, the rest will be discarded.

Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js:52]: 99% of the time there will just be a single probe. I don't think there will more than 2 or 3. If I understand the Inspector API correctly, it's only if you have more than a single breakpoint in the same location (file+line+ column).

In any case there's a follow up task after this PR to send them all as one request in an array,

**Review by simon-id — DISMISSED** [review]
LGTM for the Remote Config parts

**Review by Qard — APPROVED** [review]
One minor suggestion, but otherwise LGTM.

File: packages/dd-trace/src/debugger/devtools_client/index.js
Comment (watson) [reply packages/dd-trace/src/debugger/devtools_client/index.js]: Fixed 👍

**Review by simon-id — APPROVED** [review]
reapproving after small commit

File: packages/dd-trace/src/debugger/devtools_client/remote_config.js
Comment (bengl) [inline packages/dd-trace/src/debugger/devtools_client/remote_config.js:51]: This is a non-blocker, but maybe in a future PR you can make methods on `Session` that will post these messages so that we don't have to have these constant strings hanging around everywhere a `post()` happens.

**Review by bengl — APPROVED** [review]

**Pull request merged**
